crates/testing/src/block_builder/simple.rs

```diff
@@ -0,0 +1,368 @@
+            changes,
+        };
+
+        (source, task)
+    }
+}
+
+/// Configuration for `SimpleBuilder`
+pub struct SimpleBuilderConfig {
+    pub port: u16,
+}
+
+impl Default for SimpleBuilderConfig {
+    fn default() -> Self {
+        Self {
+            port: portpicker::pick_unused_port().expect("No free ports"),
+        }
+    }
+}
+
+#[async_trait]
+impl<TYPES: NodeType<Transaction = TestTransaction>> TestBuilderImplementation<TYPES>
+    for SimpleBuilderImplementation
+where
+    <TYPES as NodeType>::InstanceState: Default,
+{
+    type Config = SimpleBuilderConfig;
+
+    async fn start(
+        num_storage_nodes: usize,
+        config: Self::Config,
+        changes: HashMap<u64, BuilderChange>,
+    ) -> (Box<dyn BuilderTask<TYPES>>, Url) {
+        let url = Url::parse(&format!("http://localhost:{0}", config.port)).expect("Valid URL");
+
+        let (change_sender, change_receiver) = broadcast(128);
+        let (source, task) = Self::create(num_storage_nodes, changes, change_sender).await;
+        run_builder_source(url.clone(), change_receiver, source);
+
+        (Box::new(task), url)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SimpleBuilderSource<TYPES: NodeType> {
+    pub_key: TYPES::BuilderSignatureKey,
+    priv_key: <TYPES::BuilderSignatureKey as BuilderSignatureKey>::BuilderPrivateKey,
+    num_storage_nodes: usize,
+    #[allow(clippy::type_complexity)]
+    transactions: Arc<RwLock<HashMap<Commitment<TYPES::Transaction>, SubmittedTransaction<TYPES>>>>,
+    blocks: Arc<RwLock<HashMap<BuilderCommitment, BlockEntry<TYPES>>>>,
+    should_fail_claims: Arc<AtomicBool>,
+}
+
+#[async_trait]
+impl<TYPES: NodeType> ReadState for SimpleBuilderSource<TYPES> {
+    type State = Self;
+
+    async fn read<T>(
+        &self,
+        op: impl Send + for<'a> FnOnce(&'a Self::State) -> BoxFuture<'a, T> + 'async_trait,
+    ) -> T {
+        op(self).await
+    }
+}
+
+#[async_trait]
+impl<TYPES: NodeType> BuilderDataSource<TYPES> for SimpleBuilderSource<TYPES>
+where
+    <TYPES as NodeType>::InstanceState: Default,
+{
+    async fn available_blocks(
+        &self,
+        _for_parent: &VidCommitment,
+        _view_number: u64,
+        _sender: TYPES::SignatureKey,
+        _signature: &<TYPES::SignatureKey as SignatureKey>::PureAssembledSignatureType,
+    ) -> Result<Vec<AvailableBlockInfo<TYPES>>, BuildError> {
+        let transactions = self
+            .transactions
+            .read(|txns| {
+                Box::pin(async {
+                    txns.values()
+                        .filter(|txn| {
+                            // We want transactions that are either unclaimed, or claimed long ago
+                            // and thus probably not included, or they would've been decided on
+                            // already and removed from the queue
+                            txn.claimed
+                                .map(|claim_time| claim_time.elapsed() > Duration::from_secs(30))
```

Comment:
The magic number of 30 seconds here is a bit unclear. Why specifically 30 seconds? Can we get some clarification in a comment about this?

Reply:
Just realized this is a test file. I'm okay with leaving this as-is since we have magic numbers in other tests. Sorry about that!